### ISSUE TITLE
Consider introducing a Batch class to the LS pipeline.

### ISSUE THREAD

**Issue by guyboertje** [elastic/logstash#5525]
The Batch class should:
- encapsulate the batching mechanism.
- hold a list of the original events as dequeued.
- hold a list of new events created by filters (Split etc) that do not need acking - because they are not in the PQ.
- hold an AckBitSet of original events, the bits of which are flipped before an output attempts to send and after the output has successfully sent the event.
- hold a FailedBitSet of original and new events, to enable DLQ write.
- can be queried for the net uncancelled event list - being a union of the original and new event lists with cancelled event filters out.

**Comment by andrewvc** [comment]
I see the goal here, but I don

**Comment by andrewvc** [comment]
Some ideas:

I agree that we should have a Batch object, but I don't think it needs to have these extra bit sets and this delineation of new/old events. The batch object could just wrap an immutable `List<Event>`, which is all the pipeline needs to work on. It could also contain a few extra `bool` flags for things like the flush / shutdown signal.

My worry with these bitsets is that we'll have to check whether an event is new or old and alter this bitset after every single pipeline stage. It seems like if we're only doing this for the DLQ it would make more sense to just have the DLQ be exposed directly to the plugins. Only a plugin can decide if an event goes to the DLQ anyway. The plugin would just send the event to the DLQ instead of returning it. We could just have the batch contain a `List<Event>` of 'dead' items and flush those out after the batch is processed. If those items are partially processed events I think that's fine.

As far as ACKing goes I assume that the `Batch` object could just hold a special field for whatever references the PQ needs to deallocate the batch when complete.

In summary, maybe we could have

```
List<Event> items; // Never mutated, the original events
List<Event> dlq_items; // Gets added to by filters and stuff
Object queueMetadata; // Implementation specific metadata for the queue. For the PQ this might track references to internal objects that can now be dereferenced.
boolean flush; // Should this batch trigger a filter flush
boolean shutdown; // Should this batch trigger a shutdown
```

**Comment by guyboertje** [comment]
I agree about the flush and shutdown flags.

A BitSet or Dead Event List would be an internal detail, that is not exposed externally. I only meant one or the other to mark that we need to track this info in the Batch class.
The Dequeued Event List contents will always be acked, when and if, the batch instance is handed back to the PQ Read Client, so additional ack tracking inside the batch instance is not required.

A Generated Event List is needed as a separate list though, as we don't want to mix these up with the Dequeued Events. However, to filters and outputs the two Lists are seen as one via an Iterator.

The DLQ is not a concern of filters or outputs per se. They only need to call `batch.fail(event)`.
If DLQ is enabled then the batch instance is handed to the DLQ Write Client and it is this class that is concerned with pushing the dead events to its associated persistence mechanism. We can have DLQ implementations based on the PQ or a JSON file. Some users may not enable DLQ, some may want a JSON file for-the-record and others may want a PQ based DLQ that facilitates replay after a config adjustment (I'm thinking Cloud Logstash here).

In terms of memory, there will only one batch instance per worker inflight at any one time.

**Comment by andrewvc** [comment]
@guyboertje I'm with you on `batch.fail`. I just am saying we should just have that add the event reference to the DLQ list, rather than a bitset. 

My question is, I don't see why we need the Batch class to know about new events at all, it doesn't do anything with them. Given that, why not just have a `List<Event>` of events passed into `batch.fail(event)`. Simpler no? A bitset seems like extra complexity here for not much gain. A `List` of object pointers is probably fine, esp. since in the happy path there wouldn't be much use for it anyway.

**Comment by guyboertje** [comment]
The worker asks the PQ Read Client for a new batch. It receives a batch instance with a List of Events dequeued. Inside the PQ those Events are recorded as in-flight. The worker passes the batch to the filter function or next graph filter node. The first filter when iterating on the batch will be given the dequeued events. But if that filter also generates new events as a side effect, each new one is added to the batch - `batch.add(new_event)`. Internally we want to keep them separate from the dequeued events because the new events are not in the queue and therefore can't be acked. The next filter or output will get an Iterator that uses both Lists but excludes cancelled and failed (assuming that we agree that a downstream filter should not process an event that failed upstream - for some definition of failed) events.

I don't really care whether we use bitsets or lists internally - its an implementation detail that is unimportant to a design discussion.

Should a replayable DLQ have a Client that only puts failed dequeued events into the DLQ because the replay will regenerate the new events? Meaning that a for-the-record DLQ would put any failed events into the DLQ.

**Comment by guyboertje** [comment]
Also, one output may fail an event for some IO problem but another may not. Does this mean we want to record which plugin failed the event?

**Comment by andrewvc** [comment]
I think this is where the confusion lies. The pipeline really shouldn't work by mutating the batch as it moves through the pipeline IMHO. 

With something like the graph pipeline proposal there are multiple different lists of events maintained at different nodes, it doesn't really make sense to have a single list encapsulating the current state of processing. Its too tight a coupling IMHO.

Now, you could force that model to constantly reconcile itself with the Batch but I can't see to what end. Why not just have the batch contain an immutable list of the original events, and then also have a separate mutable list of failed events that can be appended to?

**Comment by jordansissel** [comment]
It is uncommon for filters to generate new events. It is also uncommon for filters to cancel (drop) events. These are weird special cases:

* Split filter creates new events and I think cancels the original event.
* Multiline filter cancels events, creates a new merged event. Can create a new event out-of-band if there's a timeout on a stream identity.
* Metrics filter creates new events out-of-band of normal event flow
* Clone creates new events.

There may be others, but for the majority of filters, the general behavior is that the event passes through, so it would be true that an immutable list is a good choice for a container for the general case. For the special cases (dropping events, injecting new events in-band, injecting new events out-of-band), maybe we can design a way to handle those separately?

**Comment by jordansissel** [comment]
In a way, I wonder if we can have:
* Immutable List<Event> for the batch of events to process.
* Filtering could return a List<Event>. 
  * In the common case where the list membership doesn't change (no drops, no new events), Filtering simply returns the original immutable list.
  * In the uncommon case of drop or new events, filtering can return a new immutable list with the dropped events removed and the new events added.
  * In the uncommon case of filters creating events out-of-band, I am not sure how best to solve this given the opportunity presented in this ticket.
* Outputs would be given an immutable List of immutable Events (cool?)

**Comment by guyboertje** [comment]
Another way to think of this is in MVC terms, where the batch is a DB, the pipeline is the Controller, a normal node is a View and a predicate node the Where clause to query over with resultant Recordset being presented to a normal node via an iterator/enumerator.

The MVC pattern exists as a decoupling mechanism and a separation of concerns.

This analogy of a batch as a DB extends to thinking of original and generated events as Tables with "cancelled", "failed", "ackable", "generated" and "current selection" as Indexes with "add" as an Insert, with "cancel" or "fail" as Updates that change an Index.

We should not see the batch as a special array or a hash of arrays, it represents more than that.

**Comment by guyboertje** [comment]
@andrewvc
The pipeline does not mutate the batch, it is the individual plugins that do that using the batch API. The pipeline should be seen as an orchestrator that selects the next plugin and brings it and the batch together. For a graph processing model, one worker thread has orchestration that walks a batch though the graph.
Each plugin should be stateless. The worker thread can only be working with one node at a time. The plugin will ask the batch for an iterator that yields events from the current selection. The plugin does not care what that selection is. The orchestration will set the batch's current selection (like a Controller does before handing off to the View) and it does this by getting a predicate and telling the batch to use the predicate to adjust its selection. This is **decouples** the plugin and pipeline from the internal batch logic.
When going down any one path in the graph, if multiple predicate nodes are seen then their effect on the selection is cumulative and when returning up the path, predicates are discarded as predicate nodes are traversed. When returning to a predicate node that has an else branch the inverse of the predicate is applied to the current selection before the else branch is walked. In my opinion, these views: "cancelled", "failed" and "current selection" are best recorded as bitsets.

**Comment by guyboertje** [comment]
@whack
Unless we continue to enforce a single worker when the multiline filter and its ilk is used, we must rule out the usage of any state-full filters in the post PQ processing. I have a plan to allow the use of state-full filters in the input Milling design.
Only if plugins have internal threads that generate events, can these be considered out-of-band because its unlikely that a batch will be around to add the generated events to - if they generate events in the worker thread then a batch is available and they are in-band. Out-of-band can be overcome by accumulating these events and adding them to a batch when next in-band call comes along.

Immutability: any failures, cancellations or additions done while iterating should not find their way back into the iteration set i.e. the current selection view is frozen during iteration.

**Comment by colinsurprenant** [comment]
it seems to me that @guyboertje batch concept proposal seems a appropriate abstraction and would in fact work for the current pipeline implementation and any future (graph or AST walking) implementation.

**Comment by andrewvc** [comment]
@guyboertje my feeling is that this would hamstring our pipeline impl. Take the pseudocode below which encapsulates the graph pipeline. Can you show me how this code would look in your example?

```ruby
events = []
graph_nodes = []
sorted = topological_sort(graph_nodes)
sorted.each do |node|
  events_in = node.in_edges.map(&:events).flatten
  edges_to_events_map = node.execute(events_in)
  node.out_edges.each {|edge| edge.events = edges_to_events_map[edge]}
end
```ruby

**Comment by colinsurprenant** [comment]
I feel we are looking at this the wrong way around. The Batch concept is and will stay the central unit of work throughout the pipeline and this batch object, from a higher level perspective should abstract away the details related to its own finalization, for such things as ACK'ing, DLQ'ing, etc. The Batch could also include such things as per-batch metrics for example. The Batch is not just a data holder for event list(s) but also encapsulate the logic that specifically apply to it. 

The internal implementation of the Batch doesn't need to be agreed upon here IMO and I don't think we need to figure all the details of how it will work specifically with the model @andrewvc proposes above for the graph pipeline. For now this is still a conceptual proposal. I'd prefer we explore, when we are ready, the best ways to have a future NG-NG (graph or other) pipeline work with such a batch object than push back the Batch proposal here because of potential inconsistencies with a POC model that is still one of many potential implementation iterations at this point. Don't get me wrong, the graph pipeline proposal is excellent with respect to the problem space it is exploring but I'd prefer we explore solutions to how an eventual graph pipeline would work efficiently around a Batch concept than trying to dismiss the Batch object idea because it doesn't tie-in directly with the proposed graph model as it is today.

**Comment by andrewvc** [comment]
@colinsurprenant @guyboertje I think the most clarifying thing we could do here is just publish some pseudocode showing what method signatures your proposed Batch object would have, and a short piece of pseudocode running the pipeline with them.

I'm 100% with you guys on not building for something that doesn't exist yet, I'm just worried that we're creating an unnecessarily tight coupling that also violates the SRP.

**Comment by guyboertje** [comment]
@andrewvc - we don't need psuedocode - [see is this PR ](https://github.com/elastic/logstash/pull/5560) for actual code proposed for the pipeline and the synchronous queue as it is now.

**Comment by guyboertje** [comment]
To summarise:

The Batch Object is a special purpose collection - its purpose is to represent the collection of events that pass through the pipeline in any one worker thread.
In a similar way that a Ruby Array has many view enumerators on its underlying data - e.g. `#select` and `#reject`, a Batch can supply appropriate view enumerators.

With a Batch object it is simple to transfer the responsibility of **tracking** which events qualify for any view enumeration, from the pipeline or plugin to the batch object - by this I mean whether cancelled and failed events are excluded from any view enumeration and whether a predicate test should be applied to further exclude events from an enumerator.

**Comment by guyboertje** [comment]
It is possible for a view via, say `#select`, to return a new ephemeral child Batch object with less events in it to be discarded when a stage is done with it. Any child batches will be created with a reference to the mother batch so that any generated/cancelled/failed events added to the child batch will be also added to the mother batch.

For example if we have this config:
```
filter {
  if [type] == 'monit' {
    mutate { add_field => { "template" => "monit_template" } }
  } elsif [type] == 'other' {
    mutate { add_field => { "template" => "other_template" } }
  } else {
    mutate { add_field => { "template" => "no_template" } }
}
```
We can apply this pseudocode
```ruby
# done in the AST but shown here
filter_nodes = []
filter_nodes << Node.new(
    <an_if_compute_component OR nested nodes>,
    Predicate.eql.lhs("[type]").rhs("monit"))
filter_nodes << Node.new(
    <an_elsif_compute_component OR nested nodes>,
    Predicate.eql.lhs("[type]").rhs("other")
  )
filter_nodes << Node.new(<an_else_compute_component OR nested nodes>, Predicate::AlwaysTrue)

temp_batch = batch
filter_nodes.each do |node, pred|
  # pred.call(event) => true|false
  in_batch = temp_batch.select(pred)
  out_batch = temp_batch.reject(pred)
  node.accept(in_batch)
  temp_batch = out_batch
end

temp_batch = batch
output_nodes.each do |node, pred|
  temp_batch = temp_batch.select(pred) # inside select: pred.call(event) => true|false
  node.accept(temp_batch)
  temp_batch = temp_batch.reject(pred)
end
```
Suppose a batch has 5 monits + 5 others + 5 foos and there a three nodes as above:
```
Node 1 (monits)
  in_batch will be 5 monits
  out_batch will be 5 others + 5 foos
  Node 1 accepts the monits
  temp_batch is 5 others + 5 foos
Node 2 (others)
  in_batch will be 5 others
  out_batch will be 5 foos
  Node 2 accepts the others
  temp_batch is 5 foos
Node 3 (else)
  in_batch will be 5 foos
  out_batch will be empty
  Node 3 accepts the foos
  temp_batch is empty
```

**Comment by guyboertje** [comment]
I would, though, prefer to have the in_batch and out_batch slices managed internally as the predicates are applied.

**Comment by colinsurprenant** [comment]
at this point we pretty much agreed between @andrewvc @jordansissel @guyboertje and myself to move on with #5560 for the batch object implementation. I believe we can now close this issue?
